lib/task-queue/WorkerTask.js

```diff
@@ -0,0 +1,192 @@
+/* globals Task */
+"use strict";
+
+const {Storage} = require("lib/task-queue/Storage.js");
+const {Cu} = require("chrome");
+Cu.import("resource://gre/modules/Task.jsm");
+
+const TASK_NEW = "new";
+const TASK_WORKING = "working";
+const TASK_DONE = "done";
+const TASK_STATUS = [TASK_NEW, TASK_WORKING, TASK_DONE];
+
+class WorkerTask {
+  /**
+   * A generic task to be executed by workers
+   *
+   * @param {String}  pageUrl
+   *                  URL of the page the task will work with
+   * @param {String}  type
+   *                  The type of worker this task relates to
+   * @param {Integer} createdAt
+   *                  The task creation date (Default: now)
+   */
+  constructor(pageUrl, type, createdAt = Date.now()) {
+    this.id = null;
+    this.pageUrl = pageUrl;
+    this.createdAt = createdAt;
+    this.jobStartedAt = null;
+    this._status = "new";
+    this.type = type;
+  }
+
+  /**
+   * Validates and sets the task status
+   *
+   * @param {String} newStatus
+   *                 The new task status
+   */
+  set status(newStatus) {
+    if (TASK_STATUS.indexOf(newStatus) === -1) {
+      throw new Error("Invalid task status.");
+    }
+    this._status = newStatus;
+  }
+
+  /**
+   * A getter for the task status
+   *
+   * @returns {String}
+   *                   Returns the task status
+   */
+  get status() {
+    return this._status;
+  }
+
+  /**
+   * Sets this task as a started job
+   */
+  jobStarted() {
+    this.jobStartedAt = Date.now();
+    this._status = TASK_WORKING;
+  }
+
+  /**
+   * Saves the current task
+   *
+   * @returns {Promise}
+   *                    Returns a promise that resolves when the task is saved
+   */
+  save() {
+    return WorkerTask.save(this);
+  }
+
+  /**
+   * Used by JSON.stringify to serialize the Task object
+   *
+   * @return {Object}
+   *                  Returns the serializable task
+   */
+  toJSON() {
+    return {
+      id: this.id,
+      pageUrl: this.pageUrl,
+      createdAt: this.createdAt,
+      jobStartedAt: this.jobStartedAt,
+      status: this._status,
+      type: this.type,
+    };
+  }
+
+  /**
+   * Creates a new task from an Object
+   *
+   * @param {Object} obj
+   *                 The object containing the task attributes values
+   */
+  static fromObject(obj) {
+    let task = new WorkerTask(obj.pageUrl, obj.type, obj.createdAt);
+    Object.assign(task, {
+      id: obj.id,
+      jobStartedAt: obj.jobStartedAt,
+      _status: obj.status,
+    });
+    return task;
+  }
+
+  /**
+   * A static method to save a task
+   *
+   * @param {Task}      task
+   *                    The task to save
+   * @returns {Promise}
+   *                    Returns a promise that resolves when the task is saved
+   */
+  static save(task) {
```

Comment:
Hmm. Perhaps you can just rename the function, otherwise you might not easily use the `static` declaration